File: to-do-app/To-Do-Java/src/main/java/com/renan/todo/controller/TaskController.java
package com.renan.todo.controller;

import com.renan.todo.controller.filter.AuthFilter;
import com.renan.todo.controller.form.MoveTaskToListForm;
import com.renan.todo.controller.form.MoveTasksFromListToListForm;
import com.renan.todo.controller.form.TaskForm;
import com.renan.todo.controller.form.UpdateTaskForm;
import com.renan.todo.dto.ErrorDTO;
import com.renan.todo.dto.TaskDTO;
import com.renan.todo.service.BusinessException;
import com.renan.todo.service.JwtService;
import com.renan.todo.service.TaskService;
import com.renan.todo.util.MessageUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * The controller for tasks operations
 */
@RequiredArgsConstructor
@RestController
public class TaskController {

    private final TaskService taskService;
    private final JwtService jwtService;

    /**
     * Get all tasks on the list
     *
     * @param jwt - user token
     * @param id  - list id
     *
     * @return - list of tasks
     */
    @GetMapping(value = "/api/taskList/tasks/{id}", produces = MediaType.APPLICATION_JSON_VALUE)
    public List<TaskDTO> getTasksByList(
            @RequestHeader(AuthFilter.AUTH) String jwt,
            @PathVariable("id") Integer id) {
        Integer userID = jwtService.getIdUser(jwt);
        return taskService.getTasksByList(id, userID);
    }

    /**
     * Mark all tasks from a list to complete
     *
     * @param jwt - user token
     * @param id  - list id
     */
    @PutMapping(value = "/api/taskList/tasks/complete/{id}")
    public void completeTasksFromList(
            @RequestHeader(AuthFilter.AUTH) String jwt,
            @PathVariable("id") Integer id) {
        Integer userID = jwtService.getIdUser(jwt);
        taskService.completeTasksFromList(id, userID);
    }

    /**
     * Mark all tasks from a list to incomplete
     *
     * @param jwt - user token
     * @param id  - list id
     */
    @PutMapping(value = "/api/taskList/tasks/uncomplete/{id}")
    public void uncompleteTasksFromList(
            @RequestHeader(AuthFilter.AUTH) String jwt,
            @PathVariable("id") Integer id) {
        Integer userID = jwtService.getIdUser(jwt);
        taskService.incompleteTasksFromList(id, userID);
    }

    /**
     * Change tasks from a list to another
     *
     * @param jwt  - user token
     * @param form - form with data for the operation
     */
    @PutMapping(value = "/api/taskList/tasks/moveFromList",
            consumes = MediaType.APPLICATION_JSON_VALUE,
            produces = MediaType.APPLICATION_JSON_VALUE)
    public void moveTasksFromList(
            @RequestHeader(AuthFilter.AUTH) String jwt,
            @RequestBody MoveTasksFromListToListForm form) {
        Integer userID = jwtService.getIdUser(jwt);
        taskService.moveTasksFromList(form.getListIdOrigin(), form.getListIdDestiny(), userID);
    }

    /**
     * Create a new task
     *
     * @param jwt  - user token
     * @param form - form with data for the operation
     *
     * @return - task created
     */
    @PostMapping(value = "/api/task",
            consumes = MediaType.APPLICATION_JSON_VALUE,
            produces = MediaType.APPLICATION_JSON_VALUE)
    public TaskDTO createTask(
            @RequestHeader(AuthFilter.AUTH) String jwt,
            @RequestBody TaskForm form) {
        Integer userID = jwtService.getIdUser(jwt);
        return taskService.createTask(
                form.getTaskName() != null ? form.getTaskName().trim() : "",
                form.getTaskDescription() != null ? form.getTaskDescription().trim() : "",
                form.getDeadline() != null ? form.getDeadline().trim() : "",
                form.getListId(),
                userID);
    }

    /**
     * Update a task
     *
     * @param jwt  - user token
     * @param id   - task id
     * @param form - form with data for the operation
     */
    @PutMapping(value = "/api/task/{id}",
            consumes = MediaType.APPLICATION_JSON_VALUE,
            produces = MediaType.APPLICATION_JSON_VALUE)
    public void updateTask(
            @RequestHeader(AuthFilter.AUTH) String jwt,
            @PathVariable("id") Integer id,
            @RequestBody UpdateTaskForm form) {
        Integer userID = jwtService.getIdUser(jwt);
        taskService.updateTask(form.getTaskName(), form.getTaskDescription(), form.getDeadline(), id, userID);
    }

    /**
     * Delete a task
     *
     * @param jwt - user token
     * @param id  - task id
     */
    @DeleteMapping(value = "/api/task/{id}")
    public void deleteTask(
            @RequestHeader(AuthFilter.AUTH) String jwt,
            @PathVariable("id") Integer id) {
        Integer userID = jwtService.getIdUser(jwt);
        taskService.deleteTask(id, userID);
    }

    /**
     * Update a task, setting as incomplete
     *
     * @param jwt - user token
     * @param id  - task id
     */
    @PutMapping(value = "/api/task/complete/{id}")
    public void complete(
            @RequestHeader(AuthFilter.AUTH) String jwt,
            @PathVariable("id") Integer id) {
        Integer userID = jwtService.getIdUser(jwt);
        taskService.completeTask(id, userID);
    }

    /**
     * Update a task, setting as incomplete
     *
     * @param jwt - user token
     * @param id  - task id
     */
    @PutMapping(value = "/api/task/uncomplete/{id}")
    public void incompleteTask(
            @RequestHeader(AuthFilter.AUTH) String jwt,
            @PathVariable("id") Integer id) {
        Integer userID = jwtService.getIdUser(jwt);
        taskService.incompleteTask(id, userID);
    }

    /**
     * Update a task, setting a new listID to it
     *
     * @param jwt  - user token
     * @param id   - task id
     * @param form - form with data for the operation
     */
    @PutMapping(value = "/api/task/list/{id}",
            consumes = MediaType.APPLICATION_JSON_VALUE,
            produces = MediaType.APPLICATION_JSON_VALUE)
    public void moveTaskToList(
            @RequestHeader(AuthFilter.AUTH) String jwt,
            @PathVariable("id") Integer id,
            @RequestBody MoveTaskToListForm form) {
        Integer userID = jwtService.getIdUser(jwt);
        taskService.moveTaskToList(id, form.getListId(), userID);
    }

    /**
     * Handle de business exceptions and return the response entity with the status code
     *
     * @param e - the exception
     *
     * @return the response entity with the status code and error dto
     */
    @ExceptionHandler(BusinessException.class)
    public ResponseEntity<ErrorDTO> handleBusinessExceptions(BusinessException e) {
        return e.getResponseEntity();
    }

    /**
     * Handle exceptions and return the response entity with the status code
     *
     * @param e - the exception
     *
     * @return the response entity with the status code and error dto
     */
    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorDTO> handleExceptions(Exception e) {
        e.printStackTrace();
        if (e instanceof BusinessException) {
            return ((BusinessException) e).getResponseEntity();
        }
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ErrorDTO(MessageUtil.getErrorMessage()));
    }

}

File: to-do-app/to-do-kotlin/src/main/kotlin/com/renan/todo/controller/TaskController.kt
package com.renan.todo.controller

import com.renan.todo.controller.filter.AUTH
import com.renan.todo.controller.form.MoveTaskToListForm
import com.renan.todo.controller.form.MoveTasksFromListToListForm
import com.renan.todo.controller.form.TaskForm
import com.renan.todo.controller.form.UpdateTaskForm
import com.renan.todo.dto.ErrorDTO
import com.renan.todo.service.BusinessException
import com.renan.todo.service.JwtService
import com.renan.todo.service.TaskService
import com.renan.todo.util.getErrorMessage
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RestController

/**
 * The controller for tasks operations
 */
@RestController
class TaskController(val taskService: TaskService, val jwtService: JwtService) {

    /**
     * Get all tasks on the list
     */
    @GetMapping(value = ["/api/taskList/tasks/{id}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun getTasksByList(
        @RequestHeader(AUTH) jwt: String?,
        @PathVariable("id") id: Int?
    ) = taskService.getTasksByList(id, jwtService.getIdUser(jwt))

    /**
     * Mark all tasks from a list to complete
     */
    @PutMapping(value = ["/api/taskList/tasks/complete/{id}"])
    fun completeTasksFromList(
        @RequestHeader(AUTH) jwt: String?,
        @PathVariable("id") id: Int?
    ) = taskService.completeTasksFromList(id, jwtService.getIdUser(jwt))

    /**
     * Mark all tasks from a list to incomplete
     */
    @PutMapping(value = ["/api/taskList/tasks/uncomplete/{id}"])
    fun uncompleteTasksFromList(
        @RequestHeader(AUTH) jwt: String?,
        @PathVariable("id") id: Int?
    ) = taskService.incompleteTasksFromList(id, jwtService.getIdUser(jwt))

    /**
     * Change tasks from a list to another
     */
    @PutMapping(
        value = ["/api/taskList/tasks/moveFromList"],
        consumes = [MediaType.APPLICATION_JSON_VALUE],
        produces = [MediaType.APPLICATION_JSON_VALUE]
    )
    fun moveTasksFromList(
        @RequestHeader(AUTH) jwt: String?,
        @RequestBody form: MoveTasksFromListToListForm
    ) = taskService.moveTasksFromList(form.listIdOrigin, form.listIdDestiny, jwtService.getIdUser(jwt))

    /**
     * Create a new task
     */
    @PostMapping(
        value = ["/api/task"],
        consumes = [MediaType.APPLICATION_JSON_VALUE],
        produces = [MediaType.APPLICATION_JSON_VALUE]
    )
    fun createTask(
        @RequestHeader(AUTH) jwt: String?,
        @RequestBody form: TaskForm
    ) = taskService.createTask(
        form.taskName.trim { it <= ' ' },
        if (form.taskDescription != null) form.taskDescription.trim { it <= ' ' } else "",
        if (form.deadline != null) form.deadline.trim { it <= ' ' } else "",
        form.listId,
        jwtService.getIdUser(jwt))

    /**
     * Update a task
     */
    @PutMapping(
        value = ["/api/task/{id}"],
        consumes = [MediaType.APPLICATION_JSON_VALUE],
        produces = [MediaType.APPLICATION_JSON_VALUE]
    )
    fun updateTask(
        @RequestHeader(AUTH) jwt: String?,
        @PathVariable("id") id: Int?,
        @RequestBody form: UpdateTaskForm
    ) = taskService.updateTask(form.taskName, form.taskDescription, form.deadline, id, jwtService.getIdUser(jwt))

    /**
     * Delete a task
     */
    @DeleteMapping(value = ["/api/task/{id}"])
    fun deleteTask(
        @RequestHeader(AUTH) jwt: String?,
        @PathVariable("id") id: Int?
    ) = taskService.deleteTask(id, jwtService.getIdUser(jwt))

    /**
     * Update a task, setting as incomplete
     */
    @PutMapping(value = ["/api/task/complete/{id}"])
    fun complete(
        @RequestHeader(AUTH) jwt: String?,
        @PathVariable("id") id: Int?
    ) = taskService.completeTask(id, jwtService.getIdUser(jwt))

    /**
     * Update a task, setting as incomplete
     */
    @PutMapping(value = ["/api/task/uncomplete/{id}"])
    fun incompleteTask(
        @RequestHeader(AUTH) jwt: String?,
        @PathVariable("id") id: Int?
    ) = taskService.incompleteTask(id, jwtService.getIdUser(jwt))

    /**
     * Update a task, setting a new listID to it
     */
    @PutMapping(
        value = ["/api/task/list/{id}"],
        consumes = [MediaType.APPLICATION_JSON_VALUE],
        produces = [MediaType.APPLICATION_JSON_VALUE]
    )
    fun moveTaskToList(
        @RequestHeader(AUTH) jwt: String?,
        @PathVariable("id") id: Int?,
        @RequestBody form: MoveTaskToListForm
    ) = taskService.moveTaskToList(id, form.listId, jwtService.getIdUser(jwt))

    /**
     * Handle de business exceptions and return the response entity with the status code
     */
    @ExceptionHandler(BusinessException::class)
    fun handleBusinessExceptions(e: BusinessException) = e.generateResponseEntity()

    /**
     * Handle exceptions and return the response entity with the status code
     */
    @ExceptionHandler(Exception::class)
    fun handleExceptions(e: Exception): ResponseEntity<ErrorDTO> {
        e.printStackTrace()
        return if (e is BusinessException) {
            e.generateResponseEntity()
        } else {
            ResponseEntity
                .status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ErrorDTO(getErrorMessage()))
        }
    }

}